### PR TITLE
DP-28505 Replace the custom responsive table scrollbar with native one

### PR DESCRIPTION
**Pull request by ygannett**
Any PRs being created needs a changelog.txt file before being merged into dev. See: [Change Log Instructions](https://github.com/massgov/mayflower/blob/develop/docs/for-developers/changelog-instructions.md)

### Table wrapper
- Replace the existing scroll bar and button control with the native scroll bar `overflow-x: auto;`
- Set up the responsive table wrapper focusable for keyboard navigation with arrow keys
     - The wrapper is focusable only when it has overflow content to enable scroll control by keyboard
- Add a label to the responsive table container syncing with the table `caption` via `aria-labelledby`
    - Add a separate instruction for screen reader users `#sr-instructions` via `aria-labelledby`

### Table caption
- Add `caption` to the template to indicate the scroll bar to users
- Style `caption` as a table title
- Hide the scroll bar indication message when the table wrapper has no scroll bar.
- When no table title with no overflow content, `caption` is hidden, not an empty element.

## Related Issue / Ticket

- [JIRA issue](https://massgov.atlassian.net/browse/DP-28505)
- [Github issue]()

## Steps to Test

1.

## Screenshots
Use something like [licecap](http://www.cockos.com/licecap/) to capture gifs to demonstrate behaviors.

## Additional Notes:

Anything else to add?

#### Impacted Areas in Application

*

#### @TODO

*

### CODE REVIEW

**Comment by ygannett** [comment]
Backstop failure is not related to the changes in this PR:  https://output.circle-artifacts.com/output/job/d7a0c3b6-2431-4488-8431-e0ef11e85e94/artifacts/0/backstop/reports/html/index.html

File: packages/assets/scss/01-atoms/_table-responsive.scss [hunk 1]
```diff
@@ -3,24 +3,36 @@
...
-    overflow: -moz-scrollbars-none;  // Firefox
 
-    &::-webkit-scrollbar {
-      display: none;  // Safari and Chrome
-    }
+  overflow-x: auto;
+
+  @media ($bp-medium-max) {
+
+    overflow-y: clip;
```
Comment (clairesunstudio) [inline packages/assets/scss/01-atoms/_table-responsive.scss]:
```suggestion
    overflow-y: hidden;
```

No difference than hidden if no [overflow-clip-margin](https://developer.mozilla.org/en-US/docs/Web/CSS/overflow-clip-margin) is set. Just want to make sure that this is the desired behavior

**Comment by clairesunstudio** [comment]
@ygannett , I'm still seeing the legacy responsive table in Mayflower Patternlab, can you please update them to use the up-to-date responsive table styles? Please let me know if I'm missing something. 

The patterns I checked:
https://mayflower.digital.mass.gov/patternlab/b/patternlab/DP-28505_scrollbar-aria/?p=viewall-atoms-table
https://mayflower.digital.mass.gov/patternlab/b/patternlab/DP-28505_scrollbar-aria/?p=pages-information-details-responsive-table
https://mayflower.digital.mass.gov/patternlab/b/patternlab/DP-28505_scrollbar-aria/?p=pages-information-details-toc-precontent-responsive-tables

**Comment by ygannett** [comment]
@clairesunstudio I wonder that's cache.  I followed your links and saw the new scroll bar.  Here is one sample.
Screenshot 2024-01-24 at 9 31 03 AM

**Comment by clairesunstudio** [comment]
@ygannett, I see the scrollbar but on mobile the table still inverts:
Screen Shot 2024-01-24 at 11 40 16 AM

For the table patterns https://mayflower.digital.mass.gov/patternlab/b/patternlab/DP-28505_scrollbar-aria/?p=viewall-atoms-table, I still see the old versions. can you update those? I think we can clean up all the old JS and assets if they are not being used on mass.gov

File: packages/assets/scss/01-atoms/_table.scss [hunk 2]
```diff
@@ -5,6 +5,26 @@
...
+    &.hide {
+      display: none;
+    }
+
+    &__content {
+      display: inline-block;
+      width: 100%;
+    }
+
+    // TODO:  ADD STYLE FOR ONE COL LAYOUT
```
Comment (clairesunstudio) [inline packages/assets/scss/01-atoms/_table.scss]: what do you mean by one col layout?

File: packages/assets/scss/01-atoms/_table.scss [hunk 3]
```diff
@@ -5,6 +5,26 @@
...
 
+  .ma__table__caption,
+  & td > table caption {
+    color: $c-font-heading;
+    font-weight: $fonts-bold;
+    line-height: 1.3;
+    text-align: left;
+    margin-bottom: 0.8em;
+
+    &.hide {
```
Comment (clairesunstudio) [inline packages/assets/scss/01-atoms/_table.scss:16]: This seems to be reverse of what's being set for the scroll info above, see .ma__table__caption__scroll-info
Just want to make sure this is needed

File: packages/patternlab/styleguide/source/assets/js/modules/responsiveTable.js [hunk 4]
```diff
@@ -213,141 +207,32 @@ export default (function (window, document, $) {
...
   function handleTableScroll(e) {
     if (skip === 0) {
       const t = e.target;
       // No matter the element, the percentage as a decimal should be the amount of pixels scrolled
       // divided by the max amount that can be scrolled which is the scroll width minus the width.
       const scrollPercent = t.scrollLeft / (t.scrollWidth - t.offsetWidth);
-      ["ma__table--responsive__wrapper", "ma__scroll-indicator", "sticky-thead-wrapper"].map(scrollable => {
+      [
+        "ma__table--responsive__wrapper",
+        "ma__scroll-indicator",
```
Comment (clairesunstudio) [inline packages/patternlab/styleguide/source/assets/js/modules/responsiveTable.js]:
```suggestion
        "ma__scroll-indicator",
```

Do we still use this class?

File: packages/patternlab/styleguide/source/assets/js/modules/responsiveTable.js [hunk 5]
```diff
@@ -213,141 +207,32 @@ export default (function (window, document, $) {
...
+        "sticky-thead-wrapper",
+      ].map((scrollable) => {
         if (t.className !== scrollable) {
           const elements = this.getElementsByClassName(scrollable);
           if (elements.length > 0) {
             let el = elements[0];
             // If we are scrolling the scrollbar or the target is the scrollbar, inverse the scroll.
-            if (scrollable === "ma__scroll-indicator" || t.className === "ma__scroll-indicator") {
+            if (
+              scrollable === "ma__scroll-indicator" ||
```
Comment (clairesunstudio) [inline packages/patternlab/styleguide/source/assets/js/modules/responsiveTable.js]: are these still relevant?

File: packages/patternlab/styleguide/source/_patterns/01-atoms/10-table/table-responsive.twig [hunk 6]
```diff
@@ -6,21 +10,12 @@
...
-    <button class="ma__table__horizontal-nav__right" aria-controls="{{tableID}}">
-      <span class="visually-hidden">Scroll right</span>
-    </button>
-  </nav>
-  <div class="ma__table--responsive__wrapper {{ mergedCells ? 'ma__table--merged-cells' : ''}}" id="{{tableID}}" tabindex="0" role="region" aria-orientation="horizontal" aria-label="{{table.description}}">
+  <div class="ma__table--responsive__wrapper {{ mergedCells ? 'ma__table--merged-cells' : ''}}" id="{{tableID}}" role="group" tabindex="-1">
     <table class="ma__table {{ tableWideClass }}">
+      <caption id="tbl-{{tableID}}" class="ma__table__caption">
+        {% if table.description %}<span class="ma__table__caption__content">{{ table.description }}</span>{% endif %}
+        <span class="ma__table__caption__scroll-info"> (Table in a horizontal scrolling container)</span>
```
Comment (clairesunstudio) [inline packages/patternlab/styleguide/source/_patterns/01-atoms/10-table/table-responsive.twig]:
```suggestion
        <span class="ma__table__caption__scroll-info"> Note: Table has hidden columns, scroll horizontally to see more.</span>

```

I think we should use more user-friendly language. Maybe something like:
"Table has hidden columns, scroll horizontally to see more."

**Comment by clairesunstudio** [comment]
Ideally, the description and scrolling instruction should not scroll with the table. But since they are in the caption, they are currently inside of the table. And whether we determine to scroll or not it's based on the table width and the table wrapper width.

I think we should probably check the tbody width and/or the thead width instead of the whole table. So that: 
1. It will prevent the description and instruction from scrolling 
2. In case the description is long, and the table itself actually don't have overflow, I think currently it will trigger the overflow instruction

**Comment by clairesunstudio** [comment]
Please resolve the conflicts on this PR and the openmass PR. Thank you!

File: packages/assets/scss/01-atoms/_table.scss (hunk 2 above)
Comment (ygannett) [reply packages/assets/scss/01-atoms/_table.scss]: This is what I forgot to remove. I initially misunderstood the task as implementing the Mayflower in openmass.  It was a note for myself.  It's removed.

File: packages/assets/scss/01-atoms/_table.scss (hunk 3 above)
Comment (ygannett) [reply packages/assets/scss/01-atoms/_table.scss:16]: The class is used by JS to hide `caption` when the table has no overflow content and no author input `caption`.
L. 133 responsiveTable.js

File: packages/patternlab/styleguide/source/assets/js/modules/responsiveTable.js (hunk 4 above)
Comment (ygannett) [reply packages/patternlab/styleguide/source/assets/js/modules/responsiveTable.js]: Thank you for mentioning that.  This is not in use and the class and related lines are cleaned up.

File: packages/patternlab/styleguide/source/assets/js/modules/responsiveTable.js (hunk 5 above)
Comment (ygannett) [reply packages/patternlab/styleguide/source/assets/js/modules/responsiveTable.js]: Not anymore.  The section has been cleaned up.

**Comment by ygannett** [comment]
I see your point.  The reason I put the scrollbar info in `caption` is that it'd be the most reasonable place as the content flow for the info for screen reader users since they know they are in the table and haven't reached the table content although that is not the best presentation for sighted users.  Another thing to consider is which is better in the limited space, especially in mobile: 1. see the table title and the scroll info. all the time by cutting off the space to show more table rows at once, 2. show the table content as much as possible by hiding the table title and the scroll info.  It depends what we put more priority than the other, I guess.  

I played around with `caption`, `table`, and `tbody`.  `caption` always stays within the `table`/`tbody` or `thead` width.  The third screenshot shows that the longer caption than the table content doesn't go over the table content (`tbody` and `thead`) width.  Instead, it gets wrapped within the table content width.  The current implementation wraps the longer caption content than the table width to fit the table width.

Table width: 100%
Caption width: 30%
Screenshot 2024-03-12 at 2 46 46 PM

Table width: 100%
Caption width: 100%
Screenshot 2024-03-12 at 2 47 19 PM

Table width: 80%
Caption width: --
Screenshot 2024-03-12 at 2 48 53 PM

**Comment by ygannett** [comment]
The backstop failures are not related to this PR's change.

File: packages/assets/scss/01-atoms/_table.scss (hunk 3 above)
Comment (clairesunstudio) [reply packages/assets/scss/01-atoms/_table.scss:16]: ah I see, so the scroll info is controlled by the .show class and the title and scroll info are both controlled by .hide

Thanks for clarifying

**Comment by clairesunstudio** [comment]
Thanks for the clear demo! So the caption length doesn't determine the table width, only the table data does. 

To your point here: 
> Another thing to consider is which is better in the limited space, especially in mobile: 1. see the table title and the scroll info. all the time by cutting off the space to show more table rows at once, 2. show the table content as much as possible by hiding the table title and the scroll info. It depends what we put more priority than the other, I guess.

I just want to clarify that I meant horizontal scrolling not vertical scrolling, just so we are on the same page. I was thinking we could treat the title and scroll info in a normal content flow outside the table so that they don't scroll horizontally with the table.  Just wanted to point out that I have not seen table title and description purposely layout like that. But if you think it's the right place for that info and the most accessible content flow,  I'm ok with that.

**Review by clairesunstudio — APPROVED** [review]

**Review by arthurbaghdas — APPROVED** [review]

**Pull request merged**